metashare/api/jobs.py

```diff
@@ -20,28 +20,59 @@
 logger = logging.getLogger(__name__)
 
 
-async def report_scratch_org_error(instance, message):
+async def report_scratch_org_error(instance, message, type_):
     from .serializers import ScratchOrgSerializer
 
     message = {
-        "type": "SCRATCH_ORG_PROVISION_FAILED",
+        "type": type_,
         "payload": {"message": message, "model": ScratchOrgSerializer(instance).data},
     }
     await push_message_about_instance(instance, message)
 
 
 @contextlib.contextmanager
-def report_errors_on(scratch_org):
+def mark_refreshing_changes(scratch_org):
+    scratch_org.currently_refreshing_changes = True
+    scratch_org.save()
+    try:
+        yield
+    finally:
+        scratch_org.currently_refreshing_changes = False
+        scratch_org.save()
+
+
+@contextlib.contextmanager
+def report_errors_on(scratch_org, type_):
     try:
         yield
     except Exception as e:
-        async_to_sync(report_scratch_org_error)(scratch_org, str(e))
+        async_to_sync(report_scratch_org_error)(scratch_org, str(e), type_)
         tb = traceback.format_exc()
         logger.error(tb)
+        raise
+
+
+@contextlib.contextmanager
+def report_errors_on_provision(scratch_org):
+    try:
+        with report_errors_on(scratch_org, "SCRATCH_ORG_PROVISION_FAILED"):
+            yield
+    except Exception:
         scratch_org.delete()
         raise
 
 
+@contextlib.contextmanager
+def report_errors_on_delete(scratch_org):
+    try:
+        with report_errors_on(scratch_org, "SCRATCH_ORG_DELETE_FAILED"):
+            yield
+    except Exception:
+        scratch_org.delete_queued_at = None
+        scratch_org.save()
```

Comment:
It looks like this is actually happening _after_ the org is serialized for the `SCRATCH_ORG_DELETE_FAILED` event. That's okay if it's better/easier that way, or necessary -- just let me know because I'll need to adjust the JS to manually remove the `delete_queued_at` if that's the case.

Reply:
It's a little easier to decompose that way, but I think you're right that it's wrong to send it that way, so I'll restructure this to fix that.